pymedphys/tests/dicom/test_anonymise.py

```diff
@@ -169,6 +169,27 @@ def test_anonymised_dataset_with_empty_patient_sex():
     assert hardcode_replace_ds["PatientSex"].value is None
 
 
+@pytest.mark.slow
+@pytest.mark.pydicom
+def test_alternative_identifying_keywords():
+    default_keyword_list = list(IDENTIFYING_KEYWORDS)
+    alternative_keyword_list = list(default_keyword_list)
+    alternative_keyword_list.append("SOPInstanceUID")
+    test_file_path = get_treatmentrecord_test_file_path()
+    ds_test = pydicom.dcmread(test_file_path, force=True)
+    anon_private_filepath = anonymise_file(
+        test_file_path,
+        delete_private_tags=True,
+        alternative_identifying_keywords=alternative_keyword_list,
+    )
+    ds_anon = pydicom.dcmread(anon_private_filepath, force=True)
+    try:
+        assert is_anonymised_file(anon_private_filepath, ignore_private_tags=False)
```

Comment:
Is it worth having `is_anonymised_file` also accept `identifying_keywords`?

Reply:
This implies considerable variability in what could be considered an anonymised file.
Which already has a (no offense) brittle definition.
So... yes, it's worthwhile, but I think it also needs to take in to account the strategy involved in anonymisation, which would then involve having an alternative (matching) strategy for determining if something was anonymised according to said particular strategy.
Or coming up with a very different definition of what "anonymised" really is.  For example by having the "non-anonymised" data in hand and then comparing to ensure that there is no identifying information.
Bottom line, I don't think it's quite the right time to refactor (redesign) is_anonymised_*().

Reply:
Yup, fair :). I'm also not opposed to its removal one day if it is too brittle.

Also, for any case within PyMedPhys, no offense needed, although we do pour ourselves and our ideas and take pride in work well done, at the end of the day it's a great thing to look at work once done, recognise it could have been done better, and change.

Reply:
Well said.
The implementation for this (in PyMedPhys) is not brittle. 
What constitutes "anonymised" is very difficult to define.
There have been a variety of papers written (I participated in/co-authored one), and a DICOM supplement was created to help address the issues, so it's a tough nut to crack.
I think that without having a set of identifying information to measure against, determining if an arbitrary DICOM object contains identifying information is a formidable task (imagine if someone's name was actually "Anonymous" :-) 

I'm a big fan of continuous improvement.
I consider the perfect to be the enemy of the good.  Good, working code provides value.  Net Present Value.
Having achieved that, and learning along the way allows for improving code.
Also, the requirements for a body of code will almost certainly change over time (typically the requirements grow).
Trying to anticipate that too much and the results of taking that approach has spawned a multitude of software development (and general management) approaches.

So... yes, it's worthwhile to re-design is_anonymised_*().  But I'd like to review what the purpose is for that set of functions, and separate out the intended uses before doing further design, re-factoring, etc.
Even though I've just said trying to anticipate too much is not a good thing, spending a little time thinking about a problem before attacking it is worthwhile.
...
Potential purposes:
- One purpose is to ensure that a particular anonymisation strategy has been implemented correctly (including various variants/flags/semaphores).
- Another purpose might be to avoid applying a strategy/anonymising (at all), if the data is already "anonymised". 
- Another purpose might be to reject data that does not meet some kind of anonymisation criteria (e.g. ensuring that PHI has been protected before uploading/transmitting data outside the appropriate domain).
- Another (perhaps nefarious) purpose might be to rapidly filter through data to determine which might be susceptible to being de-identified (albeit with a more expensive/time consuming attack). Applying this in the form of a red team/blue team approach (e.g. a GAN) could be used for beneficial purposes to strengthen anonymisation techniques.

Reply:
Yup, let's spin this off into a separate issue.